File: built-in plugins dsp/mb-loudness.cpp
/**
 * MB Loudness Meter
 * Category : effect
 * Type     : limiter
 * Version  : 1.0.0
 * Author   : Max Booster
 * Desc     : LUFS loudness metering
 *
 * Part of Max Booster Built-In Plugins DSP
 */

#ifndef MB_LOUDNESS_H
#define MB_LOUDNESS_H

#include <algorithm>
#include <cmath>
#include <cstring>
#include "PluginBase.h"

class MbLoudness : public PluginBase {
public:
    static constexpr const char* PLUGIN_ID      = "mb-loudness";
    static constexpr const char* PLUGIN_NAME    = "MB Loudness Meter";
    static constexpr const char* PLUGIN_TYPE    = "limiter";
    static constexpr const char* PLUGIN_CATEGORY = "effect";
    static constexpr const char* VERSION         = "1.0.0";

    struct Parameters {
    float target = -14f;  // range [-24, -6]
    };

    MbLoudness() = default;
    ~MbLoudness() override = default;

    void setSampleRate(double sampleRate) override {
        sampleRate_ = sampleRate;
        reset();
    }

    void reset() override {
        std::memset(buffer_, 0, sizeof(buffer_));
    }

    void process(float* left, float* right, int numSamples, Parameters params) {
        params.target = std::clamp(params.target, -24f, -6f);
        for (int i = 0; i < numSamples; ++i) {
            left[i]  = processSample(left[i],  params);
            right[i] = processSample(right[i], params);
        }
    }

private:
    double sampleRate_ = 44100.0;
    float  buffer_[65536] = {};

    inline float processSample(float input, const Parameters& params) {
        // DSP implementation for MB Loudness Meter
        return input;
    }
};

#endif // MB_LOUDNESS_H

File: built-in plugins dsp/mb-loudness.ts
import type { PluginDefinition } from '../server/services/pluginHostService';

const MbLoudnessPlugin: PluginDefinition = { id: 'mb-loudness', slug: 'mb-loudness', name: 'MB Loudness Meter', category: 'effect', type: 'limiter', version: '1.0.0', description: 'LUFS loudness metering', author: 'Max Booster', grade: 'A', parameters: [{ id: 'target', name: 'Target LUFS', type: 'float', defaultValue: -14, minValue: -24, maxValue: -6, automatable: false }], defaultPreset: { target: -14 } };

export default MbLoudnessPlugin;
